### PR TITLE
Add parse_unicode functionality

### PR DESCRIPTION
**Pull request by piratefsh**
Hi! I found the need to parse unicode values (e.g. `i like \1f36a`) instead of names (e.g `i like :cookie:`) and implemented it. I'm open to discussion on improving how this is done.

## Changes made
* Added support for unicode to img parsing with a `Twemoji.parse_unicode` function. 
* Added a few helper functions and changed some existing functions to take in an optional alternate filter function (e.g. `Twemoji.parse_document` can now take an alternate `filter_emojis` argument)
* Added tests for `Twemoji.parse_unicode`
* Fixed some duplicate test naming that was throwing warnings
* Added `Twemoji.parse_unicode` usage to README 

Cheers!
SM

### CODE REVIEW

**Comment by winston** [comment]
Just wondering from an API perspective, is there a need to expose `parse` and `parse_unicode`, or would it be a better experience to just have `parse` and let the code figure out by itself what it has to do.

In addition, this led to think about the case where a string contains both `:cookie:` and 🍪 . 

If we have two methods, this means the user would have to call `parse` then `parse_unicode` on the same string.

So maybe a better UX would be to just call `parse` and it figures out what a string contains, and does the correct conversions (meaning we do the different parses in the code).

wdyt?

**Comment by piratefsh** [comment]
That makes sense. I'm also thinking of a scenario where there might be text in the `:thing:` format that a user might not want to be parsed into emoji. I like the idea of having `parse` intelligently figure out all the emoji that needs replacing. Perhaps we can go with that and also add an option to force `:name:`-only or unicode-only parsing. How does that sound?

**Comment by winston** [comment]
> Perhaps we can go with that and also add an option to force :name:-only or unicode-only parsing. How does that sound?

Sounds good to me. It can be a second PR. @JuanitoFatas wdyt?

🙇  Thanks for your contribution!

**Comment by piratefsh** [comment]
Sure. I'd be down to work on that. You're welcome! Anything to spread emoji love 😄

**Comment by winston** [comment]
> Anything to spread emoji love 

@JuanitoFatas will be very happy to hear that. 😄

File: lib/twemoji.rb [hunk 1]
```diff
@@ -121,13 +121,50 @@ def self.parse(text, asset_root: Twemoji.configuration.asset_root,
     end
   end
 
+  # Parse string, replace emoji unicode value with image.
+  # Parse DOM, replace emoji unicode value with image.
+  #
+  # @example Usage
+  #   Twemoji.parse("I like chocolate 😍!")
```
Comment (JuanitoFatas) [inline lib/twemoji.rb]: `parse_unicode` 🙈

File: lib/twemoji.rb [hunk 2]
```diff
@@ -121,13 +121,50 @@ def self.parse(text, asset_root: Twemoji.configuration.asset_root,
...
+  #
+  # @option options [String] (optional) asset_root Asset root url to serve emoji.
+  # @option options [String] (optional) file_ext   File extension.
+  # @option options [String] (optional) class_name Emoji image's tag class attribute.
+  # @option options [String] (optional) img_attrs  Emoji image's img tag attributes.
+  #
+  # @return [String] Original text with all occurrences of emoji text
+  # replaced by emoji image according to given options.
+  def self.parse_unicode(text, asset_root: Twemoji.configuration.asset_root,
+                       file_ext:   Twemoji.configuration.file_ext,
```
Comment (JuanitoFatas) [inline lib/twemoji.rb]: would be nice if these options lines up nicely 💅

File: lib/twemoji.rb [hunk 3]
```diff
@@ -152,10 +189,12 @@ def self.options
     # @param text [String] Text string to be parse.
     # @return [String] Text with emoji text replaced by emoji image.
     # @private
-    def self.parse_html(text)
-      return text if !text.include?(":")
+    def self.parse_html(text, filter= :filter_emojis)
```
Comment (JuanitoFatas) [inline lib/twemoji.rb]: 🔎  `filter= :filter_emojis` => `filter = :filter_emojis`

File: lib/twemoji.rb [hunk 4]
```diff
@@ -152,10 +189,12 @@ def self.options
     # @param text [String] Text string to be parse.
```
Comment (JuanitoFatas) [inline lib/twemoji.rb]: 💡 Maybe we can also document filter option and its possible values

File: lib/twemoji.rb [hunk 5]
```diff
@@ -164,12 +203,12 @@ def self.parse_html(text)
     # @param doc [Nokogiri::HTML::DocumentFragment] Document to parse.
     # @return [Nokogiri::HTML::DocumentFragment] Parsed document.
     # @private
-    def self.parse_document(doc)
+    def self.parse_document(doc, filter= :filter_emojis)
```
Comment (JuanitoFatas) [inline lib/twemoji.rb]: ditto 🙈

File: lib/twemoji.rb [hunk 6]
```diff
@@ -196,8 +235,12 @@ def self.has_ancestor?(node, tags)
     # @return [String] Returns a String just like content with all emoji text
     #                  replaced by the corresponding emoji image.
     # @private
-    def self.filter_emojis(content)
-      content.gsub(emoji_pattern) { |match| img_tag(match) }
+    def self.filter_emojis(content, pattern=emoji_pattern)
```
Comment (JuanitoFatas) [inline lib/twemoji.rb]: If we want these class methods to be private, need to use `private_class_method :filter_emojis`<sup>[Source](http://ruby-doc.org/core-2.3.3/Module.html#method-i-private_class_method)</sup>

Comment (JuanitoFatas) [inline lib/twemoji.rb]: `pattern = emoji_pattern` 🙏

**Comment by JuanitoFatas** [comment]
Awesome! I left few comments :D

File: lib/twemoji.rb [hunk 7]
```diff
@@ -260,6 +306,30 @@ def self.default_attrs(name)
...
+      }
+    end
+
+
+    # Convert raw unicode to string key version
+    # e.g. 🇲🇾 converts to "1f1f2-1f1fe"
+    # @param unicode [String] Unicode value.
+    # @return String representation of unicode value
+    # @private
+    def self.unicode_to_str(unicode)
```
Comment (JuanitoFatas) [inline lib/twemoji.rb]: nice extraction here! 👏

File: test/twemoji_test.rb [hunk 8]
```diff
@@ -24,7 +23,7 @@ def test_twemoji_png
     assert_equal 1661, Twemoji.png.size
   end
 
-  def test_twemoji_png
+  def test_twemoji_svg
```
Comment (JuanitoFatas) [inline test/twemoji_test.rb]: Thanks for fixing this!!! 🐗

**Comment by JuanitoFatas** [comment]
> Perhaps we can go with that and also add an option to force :name:-only or unicode-only parsing. How does that sound?

Sounds good to me!

File: lib/twemoji.rb (hunk 6 above)
Comment (piratefsh) [inline lib/twemoji.rb]: @JuanitoFatas since `filter_emojis` is already nested under `private`, shouldn't it already be private?

**Comment by piratefsh** [comment]
Cool, thanks @JuanitoFatas! I have committed and push changes. I noticed most of it is formatting -- do you have a Ruby formatter that you used for this project?

**Comment by winston** [comment]
@piratefsh Sorry, is this ready for review?

**Comment by piratefsh** [comment]
@winston yup!

**Comment by winston** [comment]
@piratefsh ah. I thought we decided to do away with "parse_unicode" and have "parse" to be smart enough to do either?

**Comment by piratefsh** [comment]
@winston Ah, I thought you meant it can be a follow up PR after this one. I can work on that but it'll take a bit more time.

**Comment by winston** [comment]
@piratefsh oops. Sorry! I meant the "force options" can be done as a second pr. 

Take your time! No hurry!!!

**Comment by piratefsh** [comment]
@winston This is ready for review now :)

**Comment by JuanitoFatas** [comment]
This is aweeeesomeeee 💯 👏👏👏👏 !!! Thank you so much for the Pull Request 🙇 😊  I merged with some tiny changes in a493e6f.

@piratefsh Would you like to be a member of this project? 💪 🌈

**Comment by piratefsh** [comment]
@JuanitoFatas yay! thanks for the merge 😄 🎉  Sure, would be glad to contribute in any way ✨

**Pull request merged**